Introduce AudioVideoRenderer class.

jyavenard opened the pull request:
#### 6b4f3c573ce21a3d316a14591a4b2290269e96d1
<pre>
Introduce AudioVideoRenderer class.
<a href="https://bugs.webkit.org/show_bug.cgi?id=297961">https://bugs.webkit.org/show_bug.cgi?id=297961</a>
<a href="https://rdar.apple.com/159279717">rdar://159279717</a>

Reviewed by Eric Carlson.

And make MediaPlayerPrivateWebM use its AVFoundation implementation
AudioVideoRendererAVFObjC.

AudioVideoRenderer abstract the entire interaction with the system framework
required to decode and render a video or MediaSample and handle A/V synchronization.

By using an AudioVideoRenderer a MediaPlayerPrivate is only left to handle
the demuxing of the media file.

By having this new layer separation, we will be able in a future change
to have the MediaPlayerPrivate runs with its demuxer in the Content Process
while the AudioVideoRenderer runs remotely.

For now, there&apos;s no visible functional changes except on visionOS.
This current AudioVideoRendererAVFObjC is only usable on systems where
WebCoreDecompressionSession is preferred to decode MediaSample.

On visionOS the transition between an AVSampleBufferDisplayLayer and
a AVSampleBufferVideoRenderer is fully handled by AudioVideoRenderer rather
than the MediaPlayerPrivate. It no longer requires to perform internal
seeks and re-enqueuing samples.
The same improvement also occurs when the video element will be added/removed
from the DOM, the transition is instantaneous without having to perform
similar sets of operations as we no longer need to switch between an AVSBDL
and a WebCoreDecompressionSession.

Covered by existing tests.

Fullscreen, PiP and docking transitions manually tested on iPad and Vision Pro
as we don&apos;t have automated coverage for those.

* Source/WTF/wtf/LoggerHelper.h: Add DEBUG_LOG_WITH_THIS and DEBUG_LOG_WITH_THIS_IF_POSSIBLE macro.

Canonical link: <a href="https://commits.webkit.org/299373@main">https://commits.webkit.org/299373@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/020d90485350017dd5c6f8caeea5415747108379

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118794 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38475 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29126 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124982 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70856 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/998595cc-dad0-4101-a076-a9e9f71534c7) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120672 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39171 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47057 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90161 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59671 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/5c239c30-ade9-4c3f-883e-400bbae1c847) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121747 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31213 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106506 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70666 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/13ef06fd-c880-4c62-b330-c64723da1ee5) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30274 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24619 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68642 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/110912 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100656 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24807 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128033 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/117308 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45701 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34505 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98811 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46066 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102726 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98593 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25062 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44033 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22041 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42240 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45571 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51249 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/146004 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45036 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37549 "Found 1 new JSC binary failure: testapi, Found 18593 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/push2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/bug215238.shrua-2.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default ... (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48381 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46721 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/639879ca3117114dcdf5bee8d748531835db728a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49931

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/bf642a47475995237655fa0dcde44d3a28117f8b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49931

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/269bfce52effe92e0d621f9a6d1c4ee202d80c4c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49931

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/AudioVideoRendererAVFObjC.mm:
```diff
@@ -0,0 +1,1123 @@
+    }
+
+    if (type != TrackType::Audio)
+        return;
```
I'm going to change this of a switch/case instead.. nicer

eric-carlson reviewed (approved):
r=me, though another set of eyes would be helpful since this changes so much

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/AudioVideoRendererAVFObjC.mm:
```diff
@@ -0,0 +1,1141 @@
+
+void AudioVideoRendererAVFObjC::notifyWhenVideoRendererRequiresFlushToResumeDecoding(Function<void()>&& callback)
+{
+    m_videoRendererRequiresFlushToResumeDecodingCallback = WTFMove(callback);
```
There won't always be a renderer at the time this is called.
I've made the generic setup in the MediaPlayerPrivateWebM constructor; it's done once.

However, the VideoMediaSampleRenderer is lazily created only once we've started demuxing the content and processed the init segment. Only once we know there's a video track will it be created.

So this has to occur on setVideoRenderer when the VideoMediaSampleRenderer wrapper is created.

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/AudioVideoRendererAVFObjC.mm:
```diff
@@ -0,0 +1,1141 @@
+    ALWAYS_LOG(LOGIDENTIFIER);
+
+    m_sampleBufferVideoRenderer = [adoptNS(PAL::allocAVSampleBufferVideoRendererInstance()) init];
+    if (!m_sampleBufferVideoRenderer)
+        return;
```
it don't know if it can actually happen. I copied that code across. Need to look at the init code and see if it can return `nil`

jyavenard commented on Source/WebCore/platform/graphics/avfoundation/AudioVideoRendererAVFObjC.mm:
```diff
@@ -0,0 +1,1141 @@
+    m_preservePitch = preservePitch;
+    for (auto& pair : m_audioRenderers) {
+        auto& renderer = pair.second;
+        // False positive see webkit.org/b/298035
```
yeah, bad copy paste here, the line is there twice!

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a79faf5036bed5d1277c52d591d809c14d70b6b5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49931

jyavenard commented on Source/WebCore/platform/graphics/AudioVideoRenderer.h:
```diff
@@ -0,0 +1,119 @@
+    virtual void setPreservesPitch(bool) = 0;
+#if HAVE(AUDIO_OUTPUT_DEVICE_UNIQUE_ID)
+    virtual void setOutputDeviceId(const String&) { }
+#endif
```
After offline discussion with @youennf and in an attempt to coordinate our needs: I've re-organised creating various `Interfaces`, AudioInterface, VideoInterface, VideoFullscreenInterface and SynchronizerInterface and a TracksRendererManager.

For now, I've made AudioVideoRenderer inherit from all of those, so that I will be able to easily replace the AudioVideoRendererAVFObjC in the WebM player with a AudioVideoRendererRemote

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/58cbc571c061076e22536dfcdd13247210ad08b4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49931

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/62ee8f6397f7bc03d0149af09d5437d96698bb7b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49931

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/86e35ab68dbf9e5743f97ebe2c9156b994cf55ca. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49931

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8d262ea8dbb50772d7bb2c5344eae59c44239410. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49931

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/090ad05fd2009bd615cf115b081b1c709d1d2612. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49931

jyavenard committed:
Introduce AudioVideoRenderer class.
https://bugs.webkit.org/show_bug.cgi?id=297961
rdar://159279717

Reviewed by Eric Carlson.

And make MediaPlayerPrivateWebM use its AVFoundation implementation
AudioVideoRendererAVFObjC.

AudioVideoRenderer abstract the entire interaction with the system framework
required to decode and render a video or MediaSample and handle A/V synchronization.

By using an AudioVideoRenderer a MediaPlayerPrivate is only left to handle
the demuxing of the media file.

By having this new layer separation, we will be able in a future change
to have the MediaPlayerPrivate runs with its demuxer in the Content Process
while the AudioVideoRenderer runs remotely.

For now, there's no visible functional changes except on visionOS.
This current AudioVideoRendererAVFObjC is only usable on systems where
WebCoreDecompressionSession is preferred to decode MediaSample.

On visionOS the transition between an AVSampleBufferDisplayLayer and
a AVSampleBufferVideoRenderer is fully handled by AudioVideoRenderer rather
than the MediaPlayerPrivate. It no longer requires to perform internal
seeks and re-enqueuing samples.
The same improvement also occurs when the video element will be added/removed
from the DOM, the transition is instantaneous without having to perform
similar sets of operations as we no longer need to switch between an AVSBDL
and a WebCoreDecompressionSession.

Covered by existing tests.

Fullscreen, PiP and docking transitions manually tested on iPad and Vision Pro
as we don't have automated coverage for those.

* Source/WTF/wtf/LoggerHelper.h: Add DEBUG_LOG_WITH_THIS and DEBUG_LOG_WITH_THIS_IF_POSSIBLE macro.

Canonical link: https://commits.webkit.org/299373@main
